kep/61-real-load-aware-scheduling/README.md

```diff
@@ -0,0 +1,608 @@
+
+<img src="images/design.png" alt="design" width="720" height="628"/>
+
+
+
+### Metrics Provider
+
+This service provides metrics backed by a time-series database — for example, Prometheus, InfluxDB, [Kubernetes Metrics Server](https://github.com/kubernetes-sigs/metrics-server) etc.
+
+
+### Load Watcher/Load Analyser
+
+The load watcher and analyzer both run in a single process. The watcher is responsible for retrieving cluster-wide resource usage metrics like CPU, memory, network, and IO stats over windows of a specified duration from metrics providers above. It stores these in its local cache and persists several aggregations in the host local DB for fault tolerance. The analyzer is responsible for the detection of bad metrics and any remediation. Bad metrics could be due to missing metrics or metrics with considerable errors, making them anomalies. It can also be extended in the future to use ML models for analyzing metrics.
+
+Load watcher would cache metrics in the last 15-minute, 10-minute, and 5-minute windows, which can be queried via REST API exposed. A generic interface will be provided that can support fetching metrics from any metrics provider. Example is provided in Appendix.
+
+
+### DB
+
+This is a localhost database stored as a file. This DB aims to populate the load watcher's cache quickly if it is lost due to a crash and perform scheduling decisions without any impact on latency.
+The file will be stored in host file system, so it will be persisted across pod failures on the same node. For HA setups, each load watcher process will maintain its corresponding file locally.
+
+
+### Scheduler Plugins
+
+This uses the scheduler framework of K8s to incorporate our customized real load aware scheduler plugins without modifying the core scheduler code. The plugins we proposed mainly include the following two.
+
+- BestFitBinPack Plugin: It is a node sorting plugin that sorts nodes by their actual resource utilization in a way that all utilized nodes have around x% of utilization. We use x% = 50% below to describe the plugin.
+- Safe Balancing Plugin: It is a node sorting plugin that sorts nodes base on both the mean and the standard deviation of node resource utilization. It aims to balance not only the average load but also the risk caused by load variations.
+
+
+### BestFitBinPack Plugin
+
+
+#### Score plugin
+
+This plugin would extend the Score extension point. K8s scheduler framework calls the Score function for each node separately when scheduling a pod.
+
+Following is the algorithm:
+
+**Algorithm**
+
+1. Get the utilization of the current node to be scored. Call it A.
+2. Calculate the current pod's total CPU requests and overhead. Call it B.
+3. Calculate the expected utilization if the pod is scheduled under this node by adding i.e. U = A + B.
+4. If U &lt;= 50%, return U+50 as the score
+5. If 50% &lt; U &lt;= 100%, return 100 - U
```

Comment:
I still think this discontinuous function might be erratic when you have close to 50% utilization, specially given that the function uses live data. Have you experimented with other functions?

Reply:
Good point - given the nature of live data and error rate involved, the value of U should actually be 10 less than the desired utilisation value -> as per our experiments. I think I should specify this in Notes/Caveats..